### PR TITLE
[CALCITE-2257] Combination of predicates can be proved to be always true

### PR DESCRIPTION
**Pull request by vdiravka**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rex/RexCall.java [hunk 1]
```diff
@@ -128,6 +130,20 @@ public RelDataType getType() {
...
     case CAST:
       return operands.get(0).isAlwaysTrue();
+    case OR:
+      HashMap<SqlKind, RexNode> operandsMap = Maps.newHashMap();
+      for (RexNode operand : operands) {
+        SqlKind kind = operand.getKind();
+        if (operand instanceof RexCall && kind.belongsTo(SqlKind.OPPOSITE_OPERATORS)) {
+          RexCall cuurentInnerOperand = (RexCall) getOperands().get(0);
+          RexNode innerOperandOfOppositeOperator = operandsMap.get(kind.negate());
+          if (cuurentInnerOperand.equals(innerOperandOfOppositeOperator)) {
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/rex/RexCall.java]: Do you mean `currentInnerOperand`?

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml [hunk 2]
```diff
@@ -5137,4 +5137,18 @@ LogicalProject(A=[$0], B=[$1])
...
+  <TestCase name="testOrAlwaysTrue">
+    <Resource name="sql">
+      <![CDATA[select SAL from EMPNULLABLES_20
+where SAL IS NOT NULL OR SAL is null]]>
+    </Resource>
+    <Resource name="plan">
+      <![CDATA[
+LogicalProject(SAL=[$5])
+  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], SLACKER=[$8])
+    LogicalFilter(condition=[AND(=($7, 20), >($5, 1000))])
```
Comment (hsyuan) [inline core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml]: I am new to this. Can you explain to me where does the filter come from the query?

Comment (vdiravka) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml]: This is a view. You can find more details about it here:
https://github.com/apache/calcite/blob/master/core/src/test/java/org/apache/calcite/test/MockCatalogReader.java#L404

Comment (hsyuan) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml]: Got it, thanks!

**Comment by vlsi** [comment]
@vdiravka , #776 might be relevant as well.

**Comment by vdiravka** [comment]
@vlsi Does the test case from this PR pass with #776 changes?

**Comment by vlsi** [comment]
This test verifies `OR(IS NOT NULL($0), IS NULL($0))`, and that optimization is out of scope of #776 yet.

However, I think the proper placement of that kind of test is `RexProgramTest`.

**Comment by kgyrtkirk** [comment]
I'm closing this because I think right now simplification is already able to handle these cases.
```
  @Test public void testSimplify1() {
    checkSimplify(or(isNotNull(vInt()), isNull(vInt())), "true");
  }
```
Please reopen if disagree!

**Pull request closed without merging**